### ISSUE TITLE
Enhance documentation about atomicity of state transitions and the included parts

### ISSUE THREAD

**Issue by pjljvandelaar** [Systems-Modeling/SysML-v2-Release#73]
Dear SST,

In a real system, many activities occur in parallel.
To prevent race conditions, atomicity of [read-modify-write](https://en.wikipedia.org/wiki/Read%E2%80%93modify%E2%80%93write) is needed.

However, in 
https://github.com/Systems-Modeling/SysML-v2-Release/blob/master/doc/2-OMG_Systems_Modeling_Language.pdf
I only found one reference to atomicity on page 192 related to Assignment Actions:
```The values are all assigned atomically, at the same time.```

Hence, given the following basic models and a challenging constraint,
I wonder how the models could be changed to satisfy the constraint.
[Note that maybe the formalization of the constraint is wrong [but that is another issue](https://github.com/Systems-Modeling/SysML-v2-Release/issues/71) ]

```sysmlv2
package 'atomic-0' {
    attribute def RadioStartingSignal;
    attribute def RadioOnSignal;
    attribute def RadioOffSignal;

    attribute def NavigationStartingSignal;
    attribute def NavigationOnSignal;
    attribute def NavigationOffSignal;

    state def VehicleStates;

    state vehicleStates : VehicleStates parallel {
        constraint singleStart 
        {
            doc /*
                electronic devices in the vehicle are not allowed to start simultaneously.
                */
    
            not (radioStates.starting and navigationStates.starting)
        }
        
        state radioStates {
            entry; then off;

            state off;
                accept RadioStartingSignal
                    then starting;

            state starting;
                accept RadioOnSignal
                    then on;

            state on;
                accept RadioOffSignal
                    then off;
        }
        
        state navigationStates {
            entry; then off;

            state off;
                accept NavigationStartingSignal
                    then starting;

            state starting;
                accept NavigationOnSignal
                    then on;

            state on;
                accept NavigationOffSignal
                    then off;
        }
    } 
}
```

If state transitions of parallel states are executed in parallel or when interruptions are allowed,
I am so far unable to model the desired behaviour correctly in SysML v2.

If a state transition is executed sequentially, and without interruptions, I assume the following is fine.
Yet I am unable to find this information in the documentation.
```
package 'atomic-1' {
    import ScalarValues::Boolean;

    attribute def RadioStartingSignal;
    attribute def RadioOnSignal;
    attribute def RadioOffSignal;

    attribute def NavigationStartingSignal;
    attribute def NavigationOnSignal;
    attribute def NavigationOffSignal;

    state def VehicleStates;

    state vehicleStates : VehicleStates parallel {
        constraint singleStart 
        {
            doc /*
                electronic devices in the vehicle are not allowed to start simultaneously
                */
    
            not (radioStates.starting and navigationStates.starting)
        }

        attribute batteryAvailable : Boolean := true;
        
        state radioStates {
            entry; then off;

            state off;
                accept RadioStartingSignal
                    if batteryAvailable 
                        do assign batteryAvailable := false
                        then starting;

            state starting;
                accept RadioOnSignal
                    do assign batteryAvailable := true
                    then on;

            state on;
                accept RadioOffSignal
                    then off;
        }
        state navigationStates {
            entry; then off;

            state off;
                accept NavigationStartingSignal
                    if batteryAvailable 
                        do assign batteryAvailable := false
                        then starting;

            state starting;
                accept NavigationOnSignal
                    do assign batteryAvailable := true
                    then on;

            state on;
                accept NavigationOffSignal
                    then off;
        }
    } 
}
```

When entry and exit actions of a state are part of a state transition, then the following two solutions are also fine.
Yet again I am unable to find this information in the documentation.
```
package 'atomic-2' {
    import ScalarValues::Boolean;

    attribute def RadioStartingSignal;
    attribute def RadioOnSignal;
    attribute def RadioOffSignal;

    attribute def NavigationStartingSignal;
    attribute def NavigationOnSignal;
    attribute def NavigationOffSignal;

    state def VehicleStates;

    state vehicleStates : VehicleStates parallel {
        constraint singleStart 
        {
            doc /*
                electronic devices in the vehicle are not allowed to start simultaneously
                */
    
            not (radioStates.starting and navigationStates.starting)
        }
        
        attribute batteryAvailable : Boolean := true;
        
        state radioStates {
            entry; then off;

            state off;
                accept RadioStartingSignal
                    if batteryAvailable 
                        then starting;

            state starting {
                entry assign batteryAvailable := false;
                exit assign batteryAvailable := true;
            }
                accept RadioOnSignal
                    then on;

            state on;
                accept RadioOffSignal
                    then off;
        }
        state navigationStates {
            entry; then off;

            state off;
                accept NavigationStartingSignal
                    if batteryAvailable 
                        then starting;

            state starting {
                entry assign batteryAvailable := false;
                exit assign batteryAvailable := true;
            }
                accept NavigationOnSignal
                    then on;

            state on;
                accept NavigationOffSignal
                    then off;
        }
    } 
}
```
and
```
package 'atomic-3' {
    import ScalarValues::Boolean;

    attribute def RadioStartingSignal;
    attribute def RadioOnSignal;
    attribute def RadioOffSignal;

    attribute def NavigationStartingSignal;
    attribute def NavigationOnSignal;
    attribute def NavigationOffSignal;

    state def VehicleStates;

    state vehicleStates : VehicleStates parallel {
        constraint singleStart 
        {
            doc /*
                electronic devices in the vehicle are not allowed to start simultaneously
                */
    
            not (radioStates.starting and navigationStates.starting)
        }
        
        attribute batteryAvailable : Boolean := true;
            
        state radioStates {
            entry; then off;

            state off  {
                exit assign batteryAvailable := false;
            }
                accept RadioStartingSignal
                    if batteryAvailable 
                        then starting;

            state starting;
                accept RadioOnSignal
                    then on;

            state on {
                entry assign batteryAvailable := true;
            }
                accept RadioOffSignal
                    then off;
        }
        state navigationStates {
            entry; then off;

            state off  {
                exit assign batteryAvailable := false;
            }
                accept NavigationStartingSignal
                    if batteryAvailable 
                        then starting;

            state starting;
                accept NavigationOnSignal
                    then on;

            state on {
                entry assign batteryAvailable := true;
            }
                accept NavigationOffSignal
                    then off;
        }
    } 
}
```
Could you please enhance the documentation such that the execution semantics of SysML v2 becomes clear and more explicit?

P.S. You are free to use these examples in the documentation.
Furthermore, I wonder which of these examples should be labelled best practice for solving this problem in SysML v2.

**Comment by seidewitz** [comment]
Please post questions like this to the [SysML v2 Release](https://groups.google.com/g/sysml-v2-release) Google Group.
